### ISSUE TITLE
2.12 Backup and restore execution followed by an upgrade to 2.14 and compaction of tables is dropping all the data in the tables; NOTE: No TTL is present in schema definition

### ISSUE THREAD

**Issue by kripasreenivasan** [yugabyte/yugabyte-db#14592]
Jira Link: [DB-3964](https://yugabyte.atlassian.net/browse/DB-3964)
### Description

- Create a cluster in 2.12.6.1-b4
- Start sample app workloads(secondaryIndex)
- Create a colocated database , tables opted in and out of colocation, tables with collation, materialized views, gin indexes, geo-partitioned table, 
- Insert data into the above schema.
- Create a backup
- Restore the above backup
- Verify data is correctly restored
- Upgrade the cluster to 2.14.2.1-b1
- Compact tables in point 3
- Verify the data in tables post upgrade and compaction
**Issue:**
Tables are empty post compaction. 

**Note:** the tables are not created with TTL.
**This issue is observed only in 2.14**

Tserver logs
```
I1021 07:14:12.016719  2293 db_impl.cc:2905] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] Manual compaction starting
I1021 07:14:12.016937  2294 db_impl.cc:3555] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] Manual compaction from level-0 to level-0 from (begin) .. (end); will stop at (end)
I1021 07:14:12.016954  2294 compaction_job.cc:1191] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] [JOB 4] Compacting 1@0 files to L0, score -1.00
I1021 07:14:12.016960  2294 compaction_job.cc:1195] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] Compaction start summary: Base version 3 Base level 0, inputs: [13(195KB)]
I1021 07:14:12.016973  2294 event_logger.cc:66] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: EVENT_LOG_v1 {"time_micros": 1666336452016966, "job": 4, "event": "compaction_started", "files_L0": [13], "score": -1, "input_data_size": 199785}
I1021 07:14:12.017040  2294 docdb_compaction_context.cc:457] DocDB compaction feed, delete_marker_retention_time: <max>, history_cutoff={ physical: 1666335552017027 }
I1021 07:14:12.018512  2294 compaction_job.cc:933] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] [JOB 4] Compacted 1@0 files to L0 => 0 bytes
I1021 07:14:12.020982  2294 compaction_job.cc:544] T eea952666d6a4c31b239abdd65a8b5bd P 02cfdc40a3d6452fa44706d546f623c1 [R]: [default] compacted to: files[0] max score 0.00, MB/sec: 130.6 rd, 0.0 wr, level 0, files in(0, 1) out(0) MB in(0.0, 0.2) out(0.0), read-write-amplify(199785.0) write-amplify(0.0) OK, records in: 6000, records dropped: 6000
```

The automation case `testupgraderestorefirst` is failing. CC: @def- the author of the case
 CC: @bmatican

**Comment by def-** [comment]
It seems to me that compaction deletes the rows since they have expired. It is strange that the rows are marked with TTL at all. I didn't see this on master in my change adding compaction to upgrade tests.

**Comment by bmatican** [comment]
@def- @kripasreenivasan do we have the logs around, from one of these failed runs?

**Comment by def-** [comment]
Yes: https://drive.google.com/file/d/1lA84r3SgJ06-ioawxX7KAm_iC8ebfm-V/view?usp=sharing

**Comment by def-** [comment]
Only happens with colocation.

**Comment by jmeehan16** [comment]
Resolved via db341d384d77982bbd6a19e4c177818dfe29467e
